File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> anw = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        Arrays.sort(candidates);
        helper(0, 0, target, candidates, anw, list);
        return anw;
    }

    private void helper(int start, 
        int sum, 
        int target, 
        int[] candidates, 
        List<List<Integer>> anw, 
        List<Integer> list) {
        if (sum == target) {
            List<Integer> temp = new ArrayList<>(list);
            anw.add(temp);
            return;
        }
        if (start >= candidates.length || sum > target) return;

        list.add(candidates[start]);
        helper(start + 1, sum + candidates[start], target, candidates, anw, list);
        list.remove(list.size() - 1);

        while (start < candidates.length - 1 && candidates[start] == candidates[start + 1]) start++;
        helper(start + 1, sum, target, candidates, anw, list);
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> curr;
        dfs(0, 0, candidates, res, curr, target);
        return res;
    }

    void dfs(int idx, int sum, vector<int>& candidates, vector<vector<int>>& res, vector<int>& curr, int target) {
        if (sum == target) {
            res.push_back(curr);
            return;
        }
        
        if (idx >= static_cast<int>(candidates.size()) || sum > target) return;

        curr.push_back(candidates[idx]);
        dfs(idx + 1, sum + candidates[idx], candidates, res, curr, target);
        curr.pop_back();
        while (idx >= 0 && idx < candidates.size() - 1 && candidates[idx] == candidates[idx + 1]) idx++;
        dfs(idx + 1, sum, candidates, res, curr, target);
    }
};